### PR TITLE
Nesting FlowControllers to allow nested conditionals in pass manager

### LINKED ISSUE #6830: Nested flow control for Passmanagers
### What is the expected enhancement?

Skimming through the pass manager code and reading the documentation (which is quite lacking) for building pass managers I do not see an option to nest flow control. The `FlowController` class is only attached to a set of passes which makes it impossible to apply an additional flow controller on a subset of those passes. For example the use case I have in mind is to have a condition in a do while loop, for example something like:

```python
while not fixed_point:
    pass_a()
    pass_b()
    check_pass()
    if not check_pass_result:
        pass_d()
    pass_e()
pass_f()
```

which at least I couldn't find a way to represent in the current pass manager structure. It would be very useful as we make the pass manager more sophisticated to have nested flow control like this. If I just missed it we should improve the documentation around this (which honestly we need regardless).

### PR DESCRIPTION
**Pull request by irajput**
### Summary

This is another approach to fix #6830 which uses nested FlowControllers (like ConditionalController) to nest conditionals within loops in PassManager.

### CODE REVIEW

**Comment by jlapeyre** [comment]
This also fixes #5772

**Comment by jlapeyre** [comment]
PR #6938 also allows in principle nested conditions. PR #6938 achieves nesting by replacing the list of passes in a pass set with a list whose members are either passes (`BasePass`) or `PassManager`s. This PR, #6962, achieves nesting by replacing the list of passes in a pass set with a list whose members are either passes of `FlowController`s.
* A nested `PassManager` in #6938 will operate with a set of properties (essentially variables) distinct from those of the enclosing `PassManager`. This requires deciding on semantics. Not addressing semantics amounts to choosing that the properties remain distinct. 
* The nested `FlowController` #6962 operates with the properties (essentially variables) of the containing `PassManager`. The default semantics (i.e. not implementing anything to change this) is probably more useful than those #6938. The `FlowController` does not need maintain a data structure recording the state the dag and passes, rather it has read/write access to the properties of the enclosing `PassManager`
* @irajput did some quick tests showing that  #6962 is more efficient than #6938. It might be a good idea to pursue this to determine 1) how robust the efficiency difference is as the PRs stand and 2) whether #6938 could easily be made more efficient.

**Comment by 1ucian0** [comment]
depends on #6832 , `on hold`

File: test/python/transpiler/test_pass_scheduler.py
Comment (jlapeyre) [inline test/python/transpiler/test_pass_scheduler.py:565]: It's hard to see what is supposed to happen here. But, it looks like `PassA_TP_NR_NP` runs during every iteration of the while loop. Is this right? If so, it doesn't seem to check whether the conditional works.

Comment (jlapeyre) [reply test/python/transpiler/test_pass_scheduler.py:565]: This has been fixed in a commit made below: 46013e0

File: test/python/transpiler/test_pass_scheduler.py
Comment (jlapeyre) [inline test/python/transpiler/test_pass_scheduler.py:630]: I don't think this tests the feature that is added in this PR. It is in line with other tests in this file. But, it doesn't belong with this PR. In particular, the conditional passes run sequentially, rather than nested-ly.

File: test/python/transpiler/test_pass_scheduler.py
Comment (jlapeyre) [inline test/python/transpiler/test_pass_scheduler.py:659]: Again, this does not test a feature introduced in this PR. There is only a single conditional here. The PR allows nesting two or more conditionals.

File: test/python/transpiler/test_pass_scheduler.py
Comment (jlapeyre) [reply test/python/transpiler/test_pass_scheduler.py]: I removed this test in a commit.

File: test/python/transpiler/test_pass_scheduler.py
Comment (jlapeyre) [reply test/python/transpiler/test_pass_scheduler.py]: I have removed this test in a commit.

**Review by jlapeyre — APPROVED** [review]

**Comment by jlapeyre** [comment]
@1ucian0 What do you think. Can we merge this ?

**Review by 1ucian0 — APPROVED** [review]
Thanks!

**Pull request merged**
